bridge-worker/server.go

```diff
@@ -0,0 +1,195 @@
+package bridgeworker
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"time"
+
+	"cosmossdk.io/log"
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/mezo-org/mezod/bridge-worker/types"
+)
+
+type Server struct {
+	logger  log.Logger
+	server  *http.Server
+	chainID *big.Int
+}
+
+func NewServer(
+	logger log.Logger,
+	port uint16,
+	chainID *big.Int,
+) *Server {
+	s := &Server{
+		logger:  logger,
+		chainID: chainID,
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("POST /submit-signature", s.submitSignature)
+
+	s.server = &http.Server{
+		Addr:              fmt.Sprintf(":%d", port),
+		Handler:           mux,
+		ReadHeaderTimeout: time.Second,
+	}
+
+	return s
+}
+
+func (s *Server) Start() {
+	s.logger.Info("http server starting", "address", s.server.Addr)
+	go func() {
+		if err := s.server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			// should we panic here? likely yes
+			s.logger.Error("server failed to start", "error", err)
+		}
+	}()
+}
+
+func (s *Server) Stop(ctx context.Context) error {
+	return s.server.Shutdown(ctx)
+}
+
+func (s *Server) submitSignature(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	req, err := readSubmitSignatureRequest(r)
+	if err != nil {
+		writeError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	if err := req.Validate(); err != nil {
+		writeError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	address, err := s.recoverAddress(req.Entry, req.Signature)
+	if err != nil {
+		writeError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	// now do call some other service which will
+	// process the asset unlock if the address is a
+	// valid signer
+	_ = address
+
+	writeSuccess(w, http.StatusAccepted)
+}
+
+func (s *Server) recoverAddress(entry *types.AssetsUnlocked, signature string) (common.Address, error) {
+	abiEncoded, err := abiEncodeAttestationWithChainID(entry, s.chainID)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	// Apply the same TextHash transformation
+	hash := accounts.TextHash(abiEncoded)
```

Comment:
This is wrong, we need to keccak256 the `abiEncoded` var before passing to `TextHash`. I think best way to ensure both ends of the signature are consistent is extracting `attestationDigestHash` to a common place as well. Let's also move it to `ethereum/bindings/portal`.